pkg/api/v20191231preview/openshiftcluster_validate_test.go

```diff
@@ -400,3 +409,116 @@ func TestValidateWorkerProfile(t *testing.T) {
 		return v.validateWorkerProfile("properties.workerProfiles[0]", &oc.Properties.WorkerProfiles[0], &oc.Properties.MasterProfile)
 	})
 }
+
+// walk recurses through each child value of a parent value v with no cycles.
+// Excepting when v is a struct, at each step it temporarily mutates v by
+// overwriting it with its zero value, calls the test function f, then restores
+// v.  It then recurses on v's children.  The mutable field is set if any parent
+// of v is marked `mutable:"true"`.
+func walk(f func(string, bool), v reflect.Value, set func(reflect.Value), path string, mutable, ignoreCase bool) {
+	if v.Kind() != reflect.Struct {
+		current := reflect.New(v.Type()).Elem()
+		current.Set(v)
+
+		if ignoreCase && v.Kind() == reflect.String {
+			set(reflect.ValueOf(strings.ToUpper(v.String())))
+			f(path, true)
+		}
+
+		set(zeroVal(v.Type()))
+		f(path, mutable)
+
+		set(current)
+	}
+
+	switch v.Kind() {
+	case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32,
+		reflect.Int64, reflect.Uint, reflect.Uint8, reflect.Uint16,
+		reflect.Uint32, reflect.Uint64, reflect.Uintptr, reflect.Float32,
+		reflect.Float64, reflect.Complex64, reflect.Complex128, reflect.String:
+
+	case reflect.Array, reflect.Slice:
+		for i := 0; i < v.Len(); i++ {
+			walk(f, v.Index(i), v.Index(i).Set, fmt.Sprintf("%s[%d]", path, i), mutable, ignoreCase)
+		}
+
+	case reflect.Interface, reflect.Ptr:
+		if v.IsNil() {
+			return
+		}
+
+		walk(f, v.Elem(), v.Elem().Set, path, mutable, ignoreCase)
+
+	case reflect.Map:
+		i := v.MapRange()
+		for i.Next() {
+			// currently we don't recurse on keys - we assume they're simple
+			walk(f, i.Value(), func(new reflect.Value) {
+				v.SetMapIndex(i.Key(), new)
+			}, fmt.Sprintf("%s[%q]", path, i.Key()), mutable, ignoreCase)
+		}
+
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			name := strings.SplitN(v.Type().Field(i).Tag.Get("json"), ",", 2)[0]
+			if name == "" {
+				name = v.Type().Field(i).Name
+			}
+
+			mut := mutable || strings.EqualFold(v.Type().Field(i).Tag.Get("mutable"), "true")
+			ic := ignoreCase || strings.EqualFold(v.Type().Field(i).Tag.Get("mutable"), "case")
+
+			subpath := path
+			if subpath != "" {
+				subpath += "."
+			}
+			subpath += name
+
+			walk(f, v.Field(i), v.Field(i).Set, subpath, mut, ic)
+		}
+
+	default:
+		panic("unexpected kind " + v.Kind().String())
+	}
+}
+
+func zeroVal(t reflect.Type) reflect.Value {
+	return reflect.New(t).Elem()
+}
+
+func TestValidateOpenShiftClusterDelta(t *testing.T) {
+	oc, mut := validOpenShiftCluster(), validOpenShiftCluster()
+
+	v := reflect.ValueOf(mut).Elem()
+
+	walk(func(path string, mutable bool) {
```

Comment:
It seems to me that here we are basicaly testing the `github.com/jim-minter/rp/pkg/util/immutable` package. Can we test the package itself with simpler test cases with known value? Like:

* Check `mutable:"case"` validation
* Check `mutable:"true"` validation
* Check validation of fields without the tag
* Check recursive validation
* etc

Then there will be no need to *unit test* `ValidateOpenShiftClusterDelta` as it only contains `immutable.Validate` call.

IMHO, `walk` looks too complex for unit tests and requires tests itself.

Reply:
I'm glad you've got the same concerns I have.

The thing that concerns me about testing solely immutable.Validate is the risk of the actual validation somehow being wrong and not having test coverage there.

I wonder whether the test regime is trying to be too smart and we should just write and maintain all the tests manually and clearly.

But then, I wonder, if we do that for the tests - is worth bothering with immutable.Validate, and maybe we keep the actual functionality dumb as well?

Perhaps we should start with the middle ground - use immutable.Validate but test manually - and reassess later?

Reply:
> IMHO, `walk` looks too complex for unit tests and requires tests itself.

Just to clarify. I don't want us to cover `walk` with tests :) I want to get rid of it by using simplier (mock) structures.

Reply:
> Perhaps we should start with the middle ground - use immutable.Validate but test manually - and reassess later?

Yeah, sounds like a good plan.